Fix buffer leaks in FCGI and H3 `HttpClientIdleTimeoutTest`

lorban opened the pull request:
`testClientIdleTimeout` and `testRequestIdleTimeout` leak a buffer because the handler's callback is never completed and they don't give the server a chance to reclaim its resources after idle timeout.

Note: the leak was only detected in FCGI because the client was not uploading any content. Adding some uploaded content uncovered an equivalent leak in H3.

See https://github.com/eclipse/jetty.project/issues/10226

joakime reviewed (dismissed):
So this leak was entirely contained within the test case.
Interesting.
Wonder if we should document how to turn on Leak tracking in our developer guide?

lorban commented:
> So this leak was entirely contained within the test case. Interesting. Wonder if we should document how to turn on Leak tracking in our developer guide?

I'm open to the idea, but the leak tracking buffer pool currently is rough on the edges, so I'm not keen on exposing it to end users as it currently is. @gregw had a similar suggestion here (https://github.com/eclipse/jetty.project/pull/10325#issuecomment-1695235810), and we agreed that, as it stands, the leak tracker has little value outside the scope of our tests.

lorban commented:
This requires more investigation as other protocols manage to release their buffers despite the callback not being completed. We, at the very least, need to understand why not FCGI.

sbordet reviewed (changes requested):
Can we fix FCGI instead of just the test?

lorban commented:
On hold until https://github.com/eclipse/jetty.project/issues/10277 is sorted.

lorban committed:
- fix HttpClientIdleTimeoutTest to wait for server's idle timeout before checking for leaks
- improve HttpClientIdleTimeoutTest by making it upload some content
- fix FCGI server leak caused by idle timeout
- fix H3 server leak caused by idle timeout

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
- fix HttpClientIdleTimeoutTest to wait for server's idle timeout before checking for leaks
- improve HttpClientIdleTimeoutTest by making it upload some content
- fix FCGI server leak caused by idle timeout
- fix H3 server leak caused by idle timeout

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@sbordet @gregw @joakime I've reworked these tests by using `Request.addFailureListener()` to detect when the server had a chance to reclaim its resources, identified and fixed the leak in FCGI and slightly improved the tests such as a similar leak became apparent in H3 which I also fixed.

This PR is finally ready for review.

sbordet commented on jetty-core/jetty-fcgi/jetty-fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/ServerFCGIConnection.java:
```diff
@@ -222,8 +229,17 @@ void parseAndFill()
+                releaseInputBuffer();
+                break;
+            }
+            if (filled == 0)
```
You need to release also here.

lorban replied:
Yes, let me fix this.

sbordet commented on jetty-core/jetty-http3/jetty-http3-common/src/main/java/org/eclipse/jetty/http3/HTTP3Stream.java:
```diff
@@ -372,6 +372,8 @@ public void onTrailer(HeadersFrame frame)
 
     public void onFailure(long error, Throwable failure)
     {
+        HTTP3StreamConnection connection = (HTTP3StreamConnection)endPoint.getConnection();
+        connection.onFailure(error, failure);
```
Don't think it's the right place for this code.
Should all be in H3StreamConnection.

lorban replied:
The problem is that `HTTP3StreamConnection` doesn't get notified in any way. But `HttpStreamOverHTTP3.onFailure()` does get called and looks like the right place to forward the failure to `HTTP3StreamConnection`.

Let me modify this.

sbordet commented on jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientIdleTimeoutTest.java:
```diff
@@ -35,25 +35,30 @@ public class HttpClientIdleTimeoutTest extends AbstractTest
-    @Tag("DisableLeakTracking:server:FCGI")
     public void testClientIdleTimeout(Transport transport) throws Exception
     {
+        long serverIdleTimeout = idleTimeout * 3;
```
Is `* 2` enough?

lorban replied:
Indeed, `* 2` is enough.

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

lorban committed:
fix missing acquire

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix missing acquire

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented on jetty-core/jetty-fcgi/jetty-fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/ServerFCGIConnection.java:
```diff
@@ -214,17 +221,22 @@ void parseAndFill()
     {
         if (LOG.isDebugEnabled())
             LOG.debug("parseAndFill {}", this);
+        acquireInputBuffer();
```
Why is the input buffer not release when parse below returns true?

lorban replied:
You're right; the buffer could be released when empty and the parser returns true.

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

lorban committed:
release the network buffer when the parser commands to stop and it is empty

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
release the network buffer when the parser commands to stop and it is empty

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
release the network buffer when the parser commands to stop and it is empty

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
release the network buffer when the parser commands to stop and it is empty

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix over-releases and add comment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix over-releases and add comment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
@lorban the OSGI CI failures look like hard failures. @janbartel any ideas to help out?

janbartel commented:
@gregw @lorban there is a new osgi test that has been added which is failing in this branch. It seems like the new test class is there, but the new xml file that should be present in the jar is not there? Could this be a build cache issue? @olamy ?

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

olamy commented:
I can't see any more issues after rebasing from `jetty-12` branch

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

lorban committed:
Merge branch 'fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks' of github.com:jetty/jetty.project into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

lorban commented:
@sbordet nudge

gregw commented on jetty-core/jetty-fcgi/jetty-fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/ServerFCGIConnection.java:
```diff
@@ -178,8 +180,12 @@ public void onFillable()
+                    // networkBuffer field exactly when the latter isn't used anymore.
                     if (parse(networkBuffer.getByteBuffer()))
-                        return;
+                        break;
```
Is there any difference between `break` and `return` here?   or is this just a style thing?

lorban replied:
It's for style: there isn't any difference between `break` and `return` here. I just aligned this code with what the other branches were doing.

But If you prefer `return`, I'll change them all to use `return` instead of `break`.

gregw commented on jetty-core/jetty-fcgi/jetty-fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/ServerFCGIConnection.java:
```diff
@@ -214,17 +225,26 @@ void parseAndFill()
+            // networkBuffer field exactly when the latter isn't used anymore.
             if (parse(networkBuffer.getByteBuffer()))
-                return;
+                break;
```
again, no difference between `break` and `return` here, so I kind of prefer `return` as I don't need to scroll down to see what happens next

lorban replied:
Same as above: for style, so I'll change `break` for `return`.

gregw commented on jetty-core/jetty-http3/jetty-http3-common/src/main/java/org/eclipse/jetty/http3/HTTP3StreamConnection.java:
```diff
@@ -56,6 +56,15 @@ public HTTP3StreamConnection(QuicStreamEndPoint endPoint, Executor executor, Byt
 
+    public void onFailure(Throwable failure)
+    {
+        if (networkBuffer != null)
+        {
+            networkBuffer.release();
+            networkBuffer = null;
+        }
```
```suggestion
        releaseInputBuffer();
```

Can you also either change the name of that method or of the field, so they are either both `network` or both `input`.

lorban replied:
You're right, this buffer should be called `inputBuffer` and the related methods named likewise. Let me clean this.

gregw reviewed (changes requested):
a few niggles

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

sbordet committed:
Using the input buffer size for the network buffer.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>
